source/language_service/src/wrapper_refactor.rs

```diff
@@ -0,0 +1,297 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Wrapper refactor code action logic: generates a zero-parameter wrapper operation
+// that supplies default / placeholder values for an existing operation's parameters.
+
+#[cfg(test)]
+mod tests;
+
+use qsc::hir::{
+    CallableKind, ItemKind, PatKind,
+    ty::{Prim, Ty},
+};
+use qsc::{
+    Span,
+    line_column::{Encoding, Range},
+};
+
+use crate::{
+    compilation::Compilation,
+    protocol::{CodeAction, CodeActionKind, TextEdit, WorkspaceEdit},
+};
+
+pub(crate) fn operation_refactors(
+    compilation: &Compilation,
+    source_name: &str,
+    span: Span,
+    encoding: Encoding,
+) -> Vec<CodeAction> {
+    let mut code_actions = Vec::new();
+    let user_unit = compilation.user_unit();
+    let package = &user_unit.package;
+    let source_map = &user_unit.sources;
+    let source = source_map
+        .find_by_name(source_name)
+        .expect("source should exist");
+    let source_span = compilation.package_span_of_source(source_name);
+
+    for (_, item) in package.items.iter() {
+        if !source_span.contains(item.span.lo) || span.intersection(&item.span).is_none() {
+            continue;
+        }
+        if let ItemKind::Callable(decl) = &item.kind {
+            if decl.kind != CallableKind::Operation || decl.input.ty == Ty::UNIT {
+                continue; // only operations with non-empty params
+            }
+
+            // Determine indentation using source-local offset (package offset minus source base).
+            let local_lo = item.span.lo - source.offset;
+            let indent = line_indentation(&source.contents, local_lo);
+            let body_indent = if indent.contains('\t') {
+                format!("{indent}\t")
+            } else {
+                format!("{indent}    ")
+            };
```

Comment:
I think we have some code similar to this from auto-import text edits... can we reuse from there? 

https://github.com/microsoft/qdk/blob/95a6a6a91f4535508fa9185a3a2f00b81efb7294/source/language_service/src/completion/text_edits.rs